### ISSUE TITLE
Any DateTimeFormatter#withLocale should be replaced with DateTimeFormatter#localizedBy

### ISSUE THREAD

**Issue by mahozad** [thymeleaf/thymeleaf#910]
This issue is related to https://github.com/thymeleaf/thymeleaf/issues/782 in Thymeleaf core.

Java 14 and earlier had a bug in `DateTimeFormatter` class which did not format numbers correctly for locales that use different number systems (such as Persian and Arabic). This, along with the related problem in Thymeleaf core, produced incorrect formatting for numbers:

1. Somehow set the locale that Thymeleaf uses to `fa` (Persian language)
    - Creating Persian locale:  `Locale locale = Locale.forLanguageTag("fa");`
2. Output a text in an element with `th:text="${#numbers.formatInteger(1234, 1)}"`
3. The result should be **۱۲۳۴** but it was **1234**

The Thymeleaf problem was resolved. Also I filed an issue [here](https://bugs.java.com/bugdatabase/view_bug?bug_id=8243162) and the Java bug was [fixed](https://bugs.java.com/bugdatabase/view_bug?bug_id=8244245) in [Java 15](https://www.oracle.com/java/technologies/javase/15-relnote-issues.html#JDK-8244245).

So as said in [one of the comments of the JDK bug](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8243162#:~:text=the%20correct%20solution%20for%20full%20localization%20should%20be), for localization to include decimal style (numbers) etc. `DateTimeFormatter#withLocale` should be replaced with `DateTimeFormatter#localizedBy`:

```Kotlin
val formatter1 = DateTimeFormatter.ofPattern("yyyy").withLocale(Locale.forLanguageTag("fa"))
val formatter2 = DateTimeFormatter.ofPattern("yyyy").localizedBy(Locale.forLanguageTag("fa"))
val date = LocalDate.now()
println(formatter1.format(date)) // Incorrectly prints 2020
println(formatter2.format(date)) // Correctly prints ۲۰۲۰
```

In summary, any `DateTimeFormatter` creation should be accompanied by **`localizedBy(locale)`**:
  - `DateTimeFormatter`#`withLocale(locale)` should be replaced with `DateTimeFormatter`#**`localizedBy(locale)`**
  - `DateTimeFormatter.ofPattern(...);` etc.  should be accompanied by `.localizedBy(locale)` like this:
```DateTimeFormatter.ofPattern(...).localizedBy(locale);```

Also note that if Java version is < 15, then the problem may still exist unless JDK developers backport the `DateTimeFormatter` fix to earlier Java versions.

**Comment by danielfernandez** [comment]
I'm moving this to thymeleaf core because this `thymeleaf-extras-java8time` module is in the process of being integrated with the core for Thymeleaf 3.1.

However, it should be noted that using `DateTimeFormatter#localizedBy(locale)` will not be possible until Thymeleaf 4, given the fact that this method was added in JDK 10, and Thymeleaf 3.1 needs to keep JDK 8 compatibility.

**Comment by mahozad** [comment]
Thank you.

**Comment by mahozad** [comment]
For Java 14 and older, can use this code:

```java
DateTimeFormatter
    .ofPattern("yyyy")
    .localizedBy(locale)
    .withDecimalStyle(DecimalStyle.of(locale)) // 👈
    .format(myDate);
```

See 
  - [What is the difference between withLocale() and localizedBy() in Java DateTimeFormatter?](https://stackoverflow.com/questions/79744284/what-is-the-difference-between-withlocale-and-localizedby-in-java-datetimefo) 
  - https://stackoverflow.com/a/79734627/8583692.
